src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,472 @@
+        match self {
+            Json::Literal(d) => encode_literal(d, buf),
+            Json::String(d) => encode_str(d, buf),
+            Json::Object(d) => encode_obj(d, buf),
+            Json::Array(d) => encode_array(d, buf),
+            Json::Double(d) => encode_double(d, buf),
+            Json::I64(d) => encode_i64(d, buf),
+        };
+        buf.len() - start_len
+    }
+
+    fn decode(code_type: u8, data: &[u8]) -> Result<Json> {
+        match code_type {
+            TYPE_CODE_LITERAL => decode_literal(data),
+            TYPE_CODE_STRING => decode_str(data),
+            TYPE_CODE_OBJECT => decode_obj(data),
+            TYPE_CODE_ARRAY => decode_array(data),
+            TYPE_CODE_DOUBLE => decode_double(data),
+            TYPE_CODE_I64 => decode_i64(data),
+            _ => Err(invalid_type!("unsupported type {:?}", code_type)),
+        }
+    }
+
+    fn get_type_code(&self) -> u8 {
+        match *self {
+            Json::Literal(_) => TYPE_CODE_LITERAL,
+            Json::String(_) => TYPE_CODE_STRING,
+            Json::Object(_) => TYPE_CODE_OBJECT,
+            Json::Array(_) => TYPE_CODE_ARRAY,
+            Json::Double(_) => TYPE_CODE_DOUBLE,
+            Json::I64(_) => TYPE_CODE_I64,
+        }
+    }
+
+    pub fn to_string(&self) -> String {
+        serde_json::to_string(self).unwrap()
+    }
+}
+
+impl Serialize for Json {
+    fn serialize<S>(&self, serializer: S) -> SResult<S::Ok, S::Error>
+        where S: Serializer
+    {
+        match *self {
+            Json::Literal(l) => {
+                match l {
+                    JSON_LITERAL_NIL => serializer.serialize_none(),
+                    JSON_LITERAL_TRUE => serializer.serialize_bool(true),
+                    _ => serializer.serialize_bool(false),
+                }
+            }
+            Json::String(ref s) => serializer.serialize_str(s),
+            Json::Object(ref obj) => {
+                let mut map = try!(serializer.serialize_map(Some(obj.len())));
+                for (k, v) in obj {
+                    try!(map.serialize_entry(k, v));
+                }
+                map.end()
+            }
+            Json::Array(ref array) => {
+                let mut tup = try!(serializer.serialize_tuple(array.len()));
+                for item in array {
+                    try!(tup.serialize_element(item));
+                }
+                tup.end()
+            }
+            Json::Double(d) => serializer.serialize_f64(d),
+            Json::I64(d) => serializer.serialize_i64(d),
+        }
+    }
+}
+
+fn encode_literal(data: u8, buf: &mut Vec<u8>) {
+    buf.push(data);
+}
+
+fn decode_literal(data: &[u8]) -> Result<Json> {
+    Ok(Json::Literal(data[0]))
+}
+
+fn encode_str(data: String, buf: &mut Vec<u8>) {
+    buf.encode_var_u64(data.len() as u64).unwrap();
+    buf.append(&mut data.into_bytes());
+}
+
+fn decode_str(mut data: &[u8]) -> Result<Json> {
+    let length = try!(data.decode_var_u64());
```

Comment:
yes. Since we definitely want to read a `u64`,  so we needn't to pass the number to read as parameter.  Similarly, we needn't return the length we read. @siddontang